### DISCUSSION TITLE
Fish shell support

### DISCUSSION THREAD

**Discussion by josephschmitt** [cashapp/hermit#66] [Ideas]
It seems like hermit doesn't work in [fish](https://fishshell.com) at all. At first I thought it was just the shell hooks, which after looking through the [hermit source code](https://github.com/cashapp/hermit/tree/master/shell) didn't seem like it'd be that difficult to add fish support to.

But it seems like none of the activation logic works in fish either. Fish syntax isn't compatible with bash/zsh syntax, so things pretty much just break. 

Are there any plans to add support for fish?

**Comment by josephschmitt** [comment]
I've tried re-directing as much of the bash code as I can to [`bass`](https://github.com/edc/bass) and I can get decently far, but it dies at the command using `unset`. I think that code is generated by the `hermit` binary, so it would require larger changes to the project.

**Comment by alecthomas** [comment]
You're correct that Fish does bot work. Unfortunately we don't know Fish at all, but we would love to support it if someone contributed to it.

**Comment by alecthomas** [comment]
If you'd like to contribute I believe the two things that are needed are:

1. An activation script (`bin/activate-hermit.fish`) that is installed [by Hermit](https://github.com/cashapp/hermit/blob/master/env.go#L53) into the environment.
2. Implement the [shell.Shell](https://github.com/cashapp/hermit/blob/master/shell/shell.go#L26) interface.

**Reply by zph** [reply]
For future hermit<>fish users, I'm experimenting with the following successfully:

- Use direnv
- Add a custom `layout hermit` directive in your ~/.direnvrc (https://github.com/zph/zph/commit/fe12348fd15226761b693c1a18c22295301be39b)
- Include `layout hermit` in a .envrc in the root of your project folder

At that point you don't need to activate or deactivate hermit, this is handled by direnv when you enter/leave the project folder.

It uses `hermit env` to get the full environment needed for hermit and then also cleans up and resets those env vars and PATH afterwards.

**Reply by zph** [reply]
The benefit of this approach for me is that I don't need to write or maintain a custom fish activation beyond those 5 lines of code.

Thanks for writing hermit and I like the manifest format and simplicity of how it sources and installs binaries.

**Reply by alecthomas** [reply]
That is super cool, thanks for letting us know. Would you be so kind as to send a PR adding this to the documentation? That would be super helpful.

**Reply by marco-m-pix4d** [reply]
@zph the commit that you mention above, that adds the custom `layout hermit` directive, disappeared. Do you still have it somewhere? If you had time to send a PR as suggested by alec it would be even greater :-) In any case, thanks for the idea.

**Comment by mightyguava** [comment]
Merry Christmas! Fish shell support was added in #425. And #426 will add completions. I'll update when the changes are released.

**Reply by marco-m-pix4d** [reply]
Thanks @mightyguava ! :-)

**Comment by mightyguava** [comment]
Released in `v0.42.0`
